Make the timestamp field configurable

simmel opened the pull request:
Make it possible to use a field, e.g. @timestamp, for metrics sent to
Graphite instead of current timestamp. Useful when backfilling or
getting nice and complete graphs.

Thanks to @mfournier for encouraging me to submit this PR.

simmel committed:
Make the timestamp field configurable

Make it possible to use a field, e.g. @timestamp, for metrics sent to
Graphite instead of current timestamp. Useful when backfilling or
getting nice and complete graphs.

electrical commented:
I like it :-) we will review it soon and hopefully get it merged fast and released.

jsvd commented on lib/logstash/outputs/graphite.rb:
```diff
@@ -103,7 +108,11 @@ def receive(event)
 
     messages = []
-    timestamp = event.sprintf("%{+%s}")
+    timestamp = if @timestamp_field.nil?
```
Maybe default the field to @timestamp instead of nil, then you won't need the if clause

simmel replied:
I'd prefer `@timestamp` being the default too, but I'm hesitant to
change the default because of principle of least surprise when people
upgrade to 1.5.

jordansissel commented on lib/logstash/outputs/graphite.rb:
```diff
@@ -103,7 +108,11 @@ def receive(event)
+    timestamp = if @timestamp_field.nil?
+      event.sprintf("%{+%s}")
+    else
+      event[@timestamp_field].to_i
```
Add validation to verify that @timestamp_field is a time object?

jordansissel commented:
Can you add tests to cover the new behavior?

simmel commented:
Is everyone OK with @jsvd's suggestion to make `@timestamp` default?

I'll look at testing tomorrow probably @jordansissel. Got any ideas? Using `generator` then `sleep` for 2 seconds and then check that the line has timestamp which is two seconds earlier? Too error prone? Suggestions?

jsvd commented:
@simmel jordan was referring to specs that assert the configuration change has the desired effects on events.

Also.. @timestamp is already the default, right? For example, running logstash with this config:

```
input { generator { count => 1 } }
filter {
  ruby {
    code => 'event["@timestamp"] = LogStash::Timestamp.new(Time.now - 333333333)'
  }
}
output { graphite { metrics => ["test", 1] } }
```

On the other side you get an old timestamp, drawn from @timestamp:

```
% nc -l 2003
test 1.0 1085582214
% ruby -e "puts Time.at 1085582214"
2004-05-26 15:36:54 +0100
```

simmel commented:
Updated code and added a test. But I couldn't get any of the tests to run but I'm probably doing something wrong. I couldn't find any information on how to run the test. What I tried:

```
$ JAVACMD=drip LOGSTASH_DEBUG=y ~/code/logstash-1.4.2/bin/logstash rspec spec/outputs/graphite_spec.rb
Using Accessor#strict_set for specs
Using Accessor#strict_set for specs
Run options: exclude {:redis=>true, :socket=>true, :performance=>true, :couchdb=>true, :elasticsearch=>true, :elasticsearch_secure=>true, :broken=>true, :export_cypher=>true, :integration=>true}

All examples were filtered out


Finished in 0.006 seconds
0 examples, 0 failures

Randomized with seed 49950

kaka:~/code/logstash-output-graphite$
```

jsvd commented:
Good catch..you're not doing anything wrong, there are a couple problems with this plugin's tests:
1. they're tagged as "socket" and, by default, they will not be run, that's why nothing happens.
2. if you force them to run using `bundle exec rspec -t socket`, they'll complain about missing plugins (generator and kv)
3. the tests aren't prepared to run out of order and --rand makes them fail

I'll try to create the necessary PRs for fixing this asap.

jsvd commented:
@simmel can you rebase against master and try again? that and also adding "s.add_development_dependency 'logstash-filter-ruby'" to the gemspec should make "bundle install && bundle exec rspec"

simmel commented:
Updated the code, but I can't seem to run rspec:

```
$ jruby -S bundle install
WARN: Unresolved specs during Gem::Specification.reset:
      ffi (>= 0)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
Resolving dependencies...
Using rake 10.4.2
[…]
Using logstash-output-graphite 0.1.4 from source at .
Using bundler 1.7.13
Your bundle is complete!
Use `bundle show [gemname]` to see where a bundled gem is installed.
$ jruby -S bundle exec rspec
bundler: command not found: rspec
Install missing gem executables with `bundle install`
$ jruby --version
jruby 1.7.16.1 (1.9.3p392) 2014-10-28 4e93f31 on Java HotSpot(TM) 64-Bit Server VM 1.8.0_25-b17 +jit [darwin-x86_64]
```

jsvd commented:
Strange..working here with rvm and jruby 1.7.17:

```
% bundle install && bundle exec rspec                                                                                       
WARN: Unresolved specs during Gem::Specification.reset:
      ffi (>= 0)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
Resolving dependencies...
Using rake 10.4.2
...
Using logstash-filter-ruby 0.1.3
Using logstash-output-graphite 0.1.4 from source at .
Using bundler 1.7.4
Your bundle is complete!
Use `bundle show [gemname]` to see where a bundled gem is installed.
bundle install  27.48s user 1.19s system 273% cpu 10.481 total
file:/Users/joaoduarte/.rvm/rubies/jruby-head/lib/jruby.jar!/jruby/kernel/kernel.rb:28 warning: unsupported exec option: close_others
Using Accessor#strict_set for specs
Run options: exclude {:redis=>true, :socket=>true, :performance=>true, :couchdb=>true, :elasticsearch=>true, :elasticsearch_secure=>true, :broken=>true, :export_cypher=>true, :integration=>true}
.........

Finished in 0.958 seconds
9 examples, 0 failures

Randomized with seed 14879

bundle exec rspec  28.95s user 1.23s system 274% cpu 11.012 total
```

simmel commented:
Hm, I'm using jruby from homebrew. Maybe that's what's wrong? Can you
try to get it working with homebrew's jruby @jsvd?

jsvd commented:
I was able to reproduce your issue w/ homebrew's jruby.

jsvd commented:
created an issue https://github.com/elasticsearch/logstash/issues/2561

jsvd commented:
nearly there, the spec is halting because  nothing is listening in port 2003 and it tries forever to get send data there. A suggested fix:

``` ruby
  describe "timestamp_field option" do
    timestamp_new = (Time.now + 3).to_i
    let(:config) do <<-CONFIG
          input {
            generator {
              count => 1
              type => "generator"
            }
          }

          filter {
            ruby {
              code => "event['timestamp_new'] = Time.at(#{timestamp_new})"
            }
          }

          output {
            graphite {
                host => "localhost"
                port => #{port}
                timestamp_field => "timestamp_new"
                metrics => [ "foo", "1" ]
                debug => true
            }
          }
    CONFIG
    end

    it "should use the configured timestamp_field" do
      lines = server.pop
      insist { lines } =~ /^foo 1.0 #{timestamp_new}\n$/
    end
  end
```

simmel commented:
Woo, translated the spec to the new syntax too and stole your idea about using the new timestamp as an int instead of a string. (on a sidenote, how would I use a string in the logstash config in a spec? I just couldn't get:

```
    timestamp_new = (Time.now + 3).to_s
    let(:config) do <<-CONFIG
[…]
              code => "require 'time'; event['timestamp_new'] = Time.parse(\"#{timestamp_new}\")"
[…]
```

to work = /)

Anyway:

```
$ bundle exec rspec
Using Accessor#strict_set for specs
Run options: exclude {:redis=>true, :socket=>true, :performance=>true, :couchdb=>true, :elasticsearch=>true, :elasticsearch_secure=>true, :broken=>true, :export_cypher=>true, :integration=>true}
..........

Finished in 1.1 seconds
10 examples, 0 failures

Randomized with seed 40979
```

jsvd commented:
Thank you very much for your work and patience @simmel http://www.ihighfive.com/
